### PR TITLE
bitmap indices for logs 

### PR DESCRIPTION
**Pull request by AskAlexSharov**
- Indices: `addr -> bitmap(blockN)` and `topic -> bitmap(blockN)` allow make complicated queries with all-optional parameters: aka "give me logs of addres X and has topic B or has topic C and between blocks D and E"
- Indices bucket size: 13GB, 100M records (80M accounts, 20M topics). 
- Some indices have low selectivity and have big size - 1Mb. Implemented "hot-cold" sharding - small hot shard (where happen most of changes) and big cold shard (where hot shard merged when it getting big enough). Cold shards have limited size also and never merge between each-other.

### CODE REVIEW

**Comment by AskAlexSharov** [comment]
Found way to compress LogIndices bucket: 12.5GB -> 5.5Gb 
- Roaring bitmap size depends on how big value you stores: so (1,2,3) and (101,102,103) will eat different amount of disk. 
- Can get min value out of bitmap: (101,102,103) -> 101(0,1,2) and store it as bitmapMin|bitmapWithReducedDigits
- Roaring bitmap has native method roaring.AddOffset(X) - which is about "add digit X to all values  stored in bitmap". 
- Also method bitmap.Min()
- Can do  newSmallerBitmap := roaring.AddOffset(biggerBitmap, biggerBitamp.Min()) 
- And store it in db bucket as: [add or topic] -> bitmapMin_u32|AddOffset(bitmap(blockN), -bitmapMin)
- Avg speed of AddOffset on our bitmaps - 500microseconds (max 1.5ms).
- Not so bad: now 1 new block processing touches ~100 topics+addresses (means updating ~100 bitmaps)

File: cmd/rpcdaemon/commands/get_receipts.go [hunk 1]
```diff
@@ -105,34 +105,142 @@ func newFilter(addresses []common.Address, topics [][]common.Hash) *Filter {
...
 		}
 
-		begin := int64(latest)
+		begin = uint32(latest)
 		if crit.FromBlock != nil {
-			begin = crit.FromBlock.Int64()
+			begin = uint32(crit.FromBlock.Uint64())
 		}
-		end := int64(latest)
+		end = uint32(latest)
```
Comment (AlexeyAkhunov) [inline cmd/rpcdaemon/commands/get_receipts.go:135]: Why do you convert it to `uint32` and then convert to `uint64` a bit later?

File: cmd/rpcdaemon/commands/get_receipts.go [hunk 2]
```diff
@@ -105,34 +105,142 @@ func newFilter(addresses []common.Address, topics [][]common.Hash) *Filter {
...
+	var addrBitmap *gocroaring.Bitmap
+	for _, addr := range crit.Addresses {
+		m, err := bitmapdb.Get(logAddrIndex, addr[:])
+		if err != nil {
+			return nil, err
+		}
+		if addrBitmap == nil {
+			addrBitmap = m
+		} else {
+			addrBitmap = gocroaring.Or(addrBitmap, m)
```
Comment (AlexeyAkhunov) [inline cmd/rpcdaemon/commands/get_receipts.go:166]: Why do you use `Or` here and `FastOr` in another place?

File: cmd/rpcdaemon/commands/get_receipts.go [hunk 3]
```diff
@@ -105,34 +105,142 @@ func newFilter(addresses []common.Address, topics [][]common.Hash) *Filter {
...
+	}
+
+	blockNToMatchBytes := make([]byte, 4)
+
+	if blockNumbers.Cardinality() == 0 {
+		return returnLogs(logs), nil
+	}
+
+	for _, blockNToMatch := range blockNumbers.ToArray() {
+		binary.BigEndian.PutUint32(blockNToMatchBytes, blockNToMatch)
```
Comment (AlexeyAkhunov) [inline cmd/rpcdaemon/commands/get_receipts.go:185]: Do you think this needs to be removed?

File: cmd/rpcdaemon/commands/get_receipts.go (hunk 2 above)
Comment (AskAlexSharov) [reply cmd/rpcdaemon/commands/get_receipts.go:134]: Or - accept 2 bitmaps, FastOr - many, for 2 it works same as Or. so, no real reason here.

File: cmd/rpcdaemon/commands/get_receipts.go (hunk 3 above)
Comment (AskAlexSharov) [reply cmd/rpcdaemon/commands/get_receipts.go]: Yes, removed

**Pull request merged**